File: src/main/paths.js
'use strict';

/**
 * Filesystem locations.
 *
 * The database lives in the per-user application data folder. It used to sit
 * in a `data` folder beside the .exe, which was easier to find but fatally
 * fragile: the Windows uninstaller deletes the whole installation directory
 * during an update, so every update destroyed the history and the backups
 * with it. Application data survives an update by design.
 *
 * The portable build is the exception. It has no installer, nothing ever
 * uninstalls it, and keeping its data beside the .exe is the entire point —
 * copy the folder and your records travel with it.
 */

const fs = require('node:fs');
const path = require('node:path');
const os = require('node:os');

const ENV_DATA_DIR = 'ATTENDANCE_DATA_DIR';

/**
 * Folder name under %APPDATA%.
 *
 * Spelled out rather than taken from app.getName(), because the NSIS
 * installer writes to this same path while rescuing data from an older
 * version and the two must agree exactly. A rename here is a rename in
 * `build/installer.nsh` too.
 */
const APP_FOLDER = 'attendance-list';

/** Set by the portable build to the folder the .exe was run from. */
const ENV_PORTABLE_DIR = 'PORTABLE_EXECUTABLE_DIR';

let resolved = null;

/** Folder the app runs from: next to the .exe when packaged, repo root in dev. */
function appDir(app) {
  if (app && app.isPackaged) {
    return path.dirname(app.getPath('exe'));
  }
  return path.resolve(__dirname, '..', '..');
}

/** Roaming application data — %APPDATA% on Windows. */
function userDataDir(app) {
  if (app) return path.join(app.getPath('appData'), APP_FOLDER);
  return path.join(os.homedir(), `.${APP_FOLDER}`);
}

function isWritable(dir) {
  try {
    fs.mkdirSync(dir, { recursive: true });
    const probe = path.join(dir, '.write-test');
    fs.writeFileSync(probe, 'ok');
    fs.unlinkSync(probe);
    return true;
  } catch {
    return false;
  }
}

/** Where a portable build keeps its data, or null when this is not one. */
function portableDir() {
  const beside = process.env[ENV_PORTABLE_DIR];
  return beside ? path.join(path.resolve(beside), 'data') : null;
}

function dataDir(app) {
  const override = process.env[ENV_DATA_DIR];
  if (override) return path.resolve(override);

  if (resolved) return resolved;

  const portable = portableDir();
  if (portable && isWritable(portable)) {
    resolved = portable;
    return resolved;
  }

  const preferred = path.join(userDataDir(app), 'data');
  resolved = isWritable(preferred) ? preferred : path.join(appDir(app), 'data');
  return resolved;
}

/**
 * Where earlier versions kept the database, newest arrangement first.
 *
 * Read on startup so that an existing installation carries its history into
 * the new location instead of opening to an empty list.
 */
function legacyDataDirs(app) {
  const dirs = [path.join(appDir(app), 'data')];
  if (app) {
    dirs.push(path.join(app.getPath('appData'), 'AttendanceList'));
    dirs.push(path.join(app.getPath('userData'), 'data'));
  }
  const current = dataDir(app);
  return dirs.filter((dir) => path.resolve(dir) !== path.resolve(current));
}

const dbPath = (app) => path.join(dataDir(app), 'attendance.db');

/** Default backup folder: somewhere the person can actually find it. */
function defaultBackupDir(app) {
  if (app) {
    try {
      return path.join(app.getPath('documents'), 'Attendance List', 'Backups');
    } catch {
      /* no Documents folder — fall through */
    }
  }
  return path.join(dataDir(app), 'backups');
}

function ensureDirs(app) {
  fs.mkdirSync(dataDir(app), { recursive: true });
}

/** True when the location was set deliberately, e.g. a shared network folder. */
const isCustom = () => Boolean(process.env[ENV_DATA_DIR]);

/**
 * True when the preferred location was not writable and we fell back.
 *
 * A deliberately configured folder is not a fallback, however unusual it
 * looks — warning about someone's own choice trains them to ignore warnings.
 */
function usingFallback(app) {
  if (isCustom()) return false;
  const preferred = portableDir() || path.join(userDataDir(app), 'data');
  return path.resolve(dataDir(app)) !== path.resolve(preferred);
}

/** True when this is the portable build. */
const isPortable = () => Boolean(portableDir());

/** Reset the memoised location. Tests only. */
function _reset() {
  resolved = null;
}

module.exports = {
  ENV_DATA_DIR,
  ENV_PORTABLE_DIR,
  APP_FOLDER,
  appDir,
  userDataDir,
  dataDir,
  dbPath,
  defaultBackupDir,
  legacyDataDirs,
  ensureDirs,
  usingFallback,
  isCustom,
  isPortable,
  _reset,
};

File: app/paths.py
"""Filesystem locations used by the application.

The database lives in a ``data`` folder beside the application itself, so
everything to do with attendance sits in one place you can see, copy or back
up by hand.

Replacing the executable during an update does not disturb that folder — the
updater swaps a single file. The one case the folder cannot be used is an
install location the user cannot write to (``C:\\Program Files``, or running
from a locked-down network share), so :func:`data_dir` probes for writability
once and falls back to the per-user application data directory.
"""

from __future__ import annotations

import os
import sys
from pathlib import Path
from typing import Optional

from .version import APP_NAME

_DIR_NAME = "AttendanceList"
_DATA_FOLDER = "data"

_resolved_data_dir: Optional[Path] = None


def is_frozen() -> bool:
    """True when running from a PyInstaller bundle rather than source."""
    return bool(getattr(sys, "frozen", False))


#: Overrides the data location. Useful for tests and for pointing an install
#: at a shared drive so several PCs read one attendance database.
ENV_DATA_DIR = "ATTENDANCE_DATA_DIR"


def app_dir() -> Path:
    """The folder the app lives in — next to the .exe, or the project root."""
    if is_frozen():
        return Path(sys.executable).resolve().parent
    return Path(__file__).resolve().parent.parent


def user_data_dir() -> Path:
    """Per-user fallback, used only when :func:`app_dir` is not writable."""
    if sys.platform == "win32":
        base = Path(os.environ.get("LOCALAPPDATA") or Path.home() / "AppData" / "Local")
    elif sys.platform == "darwin":
        base = Path.home() / "Library" / "Application Support"
    else:
        base = Path(os.environ.get("XDG_DATA_HOME") or Path.home() / ".local" / "share")
    return base / _DIR_NAME


def _is_writable(candidate: Path) -> bool:
    """Can we actually create files here? Permissions alone lie on Windows."""
    try:
        candidate.mkdir(parents=True, exist_ok=True)
        probe = candidate / ".write-test"
        probe.write_text("ok", encoding="utf-8")
        probe.unlink()
        return True
    except Exception:
        return False


def data_dir() -> Path:
    """Where the database and backups live.

    Beside the app by default; the result is resolved once and cached so the
    writability probe does not run on every call.
    """
    override = os.environ.get(ENV_DATA_DIR)
    if override:
        return Path(override).expanduser()

    global _resolved_data_dir
    if _resolved_data_dir is not None:
        return _resolved_data_dir

    beside_app = app_dir() / _DATA_FOLDER
    _resolved_data_dir = beside_app if _is_writable(beside_app) else user_data_dir()
    return _resolved_data_dir


def using_fallback_location() -> bool:
    """True when the app folder was read-only and we fell back."""
    return data_dir() != app_dir() / _DATA_FOLDER


def db_path() -> Path:
    return data_dir() / "attendance.db"


def backup_dir() -> Path:
    return data_dir() / "backups"


def default_export_dir() -> Path:
    """Where export dialogs open by default."""
    downloads = Path.home() / "Downloads"
    return downloads if downloads.is_dir() else Path.home()


def log_path() -> Path:
    return data_dir() / "app.log"


def resource_path(relative: str) -> Path:
    """Resolve a bundled read-only resource (icons, themes).

    PyInstaller unpacks one-file bundles into ``sys._MEIPASS`` at runtime.
    """
    base = Path(getattr(sys, "_MEIPASS", Path(__file__).resolve().parent.parent))
    return base / relative


def ensure_dirs() -> None:
    """Create the directories the app writes to. Safe to call repeatedly."""
    data_dir().mkdir(parents=True, exist_ok=True)
    backup_dir().mkdir(parents=True, exist_ok=True)


def adopt_legacy_database() -> Optional[Path]:
    """Move a database left in the old per-user location into the app folder.

    Earlier builds stored data under Application Support / LOCALAPPDATA. If
    one is found and the new location is still empty, bring it across so no
    attendance history is stranded. The original is renamed rather than
    deleted, so a mistake here is recoverable.
    """
    target = db_path()
    if target.exists() or os.environ.get(ENV_DATA_DIR):
        return None

    legacy = user_data_dir() / "attendance.db"
    if not legacy.exists() or legacy.parent == data_dir():
        return None

    try:
        ensure_dirs()
        # SQLite's backup API rather than a file copy: the database runs in
        # WAL mode, so recent writes may live in the -wal sidecar. Copying
        # only the .db would silently drop them.
        import sqlite3

        source = sqlite3.connect(str(legacy))
        destination = sqlite3.connect(str(target))
        try:
            source.backup(destination)
        finally:
            destination.close()
            source.close()

        legacy.rename(legacy.with_suffix(".db.migrated"))
        for suffix in ("-wal", "-shm"):
            leftover = legacy.with_name(legacy.name + suffix)
            if leftover.exists():
                leftover.unlink()
        return target
    except Exception:
        # Never let a migration problem stop the app from opening.
        return None


def describe_data_location() -> str:
    """Human-readable data location, shown in the UI so users can find it."""
    return f"{APP_NAME} data: {data_dir()}"
